### ISSUE TITLE
Change AdviceMap key in JSON inputs to use [u64; 4]

### ISSUE THREAD

**Issue by frisitano** [0xPolygonMiden/miden-vm#993]
To define entries in the advice map using a JSON input file we currently use a hexadecimal string to represent the `Word` key.  The `Felt`s in the word are encoded as little endian u64 and then the hex outputs combined to produce the final hex string.  Instead we should migrate to specifying the keys as [u64; 4] in JSON such that the keys can be easily interpreted from the JSON.

**Comment by bobbinth** [comment]
What is the difficulty of dealing with hex encoding? Is it because there is no straight-forward way to convert hex strings to/from words? If so, maybe a better solution would be provide utility functions to do something like `word_to_hex()` and `word_from_hex()`?

**Comment by frisitano** [comment]
> What is the difficulty of dealing with hex encoding? Is it because there is no straight-forward way to convert hex strings to/from words? If so, maybe a better solution would be provide utility functions to do something like `word_to_hex()` and `word_from_hex()`?

Yes this is part of the problem.  If we want to provide the utility methods then they should be provided via the CLI. I think this will help quite a bit.

However, there is also a formatting difference between how keys are referenced in the JSON input file and assembly code.  If we could achieve parity between the two then it may yield superior UX. E.g.

`inputs.json`:
```json
    "advice_map": {
        "0.0.0.1": [5, 6, 7, 8],
        "0.0.0.2": [9, 10, 11, 12],
        "0.0.0.3": [13, 14, 15, 16]
    },
```

`code.masm`:
```
...
push.0.0.0.1 adv.push_mapval 
```

In my opinion it's easier to cross reference between masm and `inputs.json` making UX better.

**Comment by bobbinth** [comment]
> However, there is also a formatting difference between how keys are referenced in the JSON input file and assembly code. If we could achieve parity between the two then it may yield superior UX.

Agreed that we should have parity between the two - but I think we can do this with hex strings as well. Specifically, [push](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/io_operations.html#constant-inputs) instruction supports immediates formatted as hex strings. For example, the following should be equivalent:
```
push.0x1234.0xabcd
push.0x0000000000001234000000000000abcd
```
And I think using hex strings in both places is more intuitive as people usually think of hashes as strings.

**Comment by frisitano** [comment]
> Agreed that we should have parity between the two - but I think we can do this with hex strings as well. Specifically, [push](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/io_operations.html#constant-inputs) instruction supports immediates formatted as hex strings. For example, the following should be equivalent:
> 
> ```
> push.0x1234.0xabcd
> push.0x0000000000001234000000000000abcd
> ```
> 
> And I think using hex strings in both places is more intuitive as people usually think of hashes as strings.

It slipped my mind that we can use hex representation with the push instruction. I agree this is preferable.

**Comment by saxenism** [comment]
If this issue is still open and relevant, I'd like to tackle this.

However, can someone (@frisitano @bobbinth) summarize what exactly is the ideal output when this issue is resolved?

From what I can see, we want parity between how keys are referenced in the JSON input file and assembly code and that parity is in the form of hex strings.

Since [push](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/io_operations.html#constant-inputs) instruction supports immediates formatted as hex strings already, what exactly needs to be done?

**Comment by frisitano** [comment]
> If this issue is still open and relevant, I'd like to tackle this.
> 
> 
> 
> However, can someone (@frisitano @bobbinth) summarize what exactly is the ideal output when this issue is resolved?
> 
> 
> 
> From what I can see, we want parity between how keys are referenced in the JSON input file and assembly code and that parity is in the form of hex strings.
> 
> 
> 
> Since [push](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/io_operations.html#constant-inputs) instruction supports immediates formatted as hex strings already, what exactly needs to be done?

Hey, thanks for you interest in contributing! I will need to review this issue in more depth but I believe, as you mentioned, we already have what is required to achieve parity.  As such, there is nothing to do on this issue.  I will take a closer look later to confirm.  In the meantime, please feel free to review other issues and let me know if there's anything else you would like to work on and I will assist in getting you started.

**Comment by Overcastan** [comment]
We have a problem with using hex strings. It can be easily shown on the example.
Assume we added a Merkle Tree to the Merkle Store. After that the user will be shown the message (currently not, but I'm working on it):
```
Added Merkle tree with root 0x0463f7d47758ad94b11dbf9675ffb7b331baa9c150d7fac6d784055c313eab0e to the Merkle store
```
where root is a hex representation of the `RpoDigest`. But if user wants to use this root to get node from the tree (or perform any other operation that uses the root), the program will fail:
```
begin 
    # push the root to the stack
    push.0x0463f7d47758ad94b11dbf9675ffb7b331baa9c150d7fac6d784055c313eab0e
    # get the value with index `0` on depth `2` from the Merkle Tree
    push.0.2
    mtree_get
end
``` 
Reason — `RootNotInStore(...)`. So the provided root does not match any root in the Merkle Store.

Here's what was discovered after testing: 
- `Display` trait for the `RpoDigest` is implemented in such way that `Felt` array is converted to the **Little-endian** bytes array and than casted to the hex string.
-  Hex string parser of the `push` operation trying to get `u64` values from the hex string using `u64::from_str_radix` function. But judging by the tests this function assumes hex to be created from the array of **the Big-endian** bytes, that's why the resulting values that are put to the stack turned out to be incorrect. Using the hex string created from an array of Big-endian bytes and than parsing it with `u64::from_str_radix` led to the correct result.

There are two main approaches: change hex parsing in the `push` operation to make the resulting values Little-endian and change `RpoDigest` display, (de)serialization and other implementation to use Big-endian system. Both of them are breaking the tests. For now first approach looks better, but we should discuss this first.

**Comment by bobbinth** [comment]
I agree that converting `push` instruction to interpret hex strings in little-endian form is probably a better approach - but curious to see what others think.

**Comment by bobbinth** [comment]
With the work done in #1076, I think we can close this. One thing left to do is to make sure the [MerkleStore](https://github.com/0xPolygonMiden/miden-vm/tree/next/miden/examples/merkle_store) example still works. @Overcastan - could you check this and update if needed?

**Comment by bobbinth** [comment]
Closed by #1076 and #1086.
